File: ChipTallyiOS/ChipTally/Domain/UseCases/PotUseCase.swift
//
//  PotUseCase.swift
//  ChipTally
//

import Foundation

protocol PotUseCaseProtocol {
    func bet(session: inout GameSession, playerId: UUID, amount: Int) -> Result<Transaction, TransferError>
    func collectPot(session: inout GameSession, winnerId: UUID) -> Result<Transaction, TransferError>
    func currentRoundBet(session: GameSession, playerId: UUID) -> Int
}

/// 보드 가운데 팟에 칩을 걸고, 승자가 전액을 가져가는 흐름.
///
/// '이번 판' 은 마지막 `.potWin` 이후를 뜻한다. 별도 상태를 두지 않고 거래 내역에서
/// 계산하므로 저장/복원과 자동으로 맞아떨어진다.
final class PotUseCase: PotUseCaseProtocol {

    func bet(session: inout GameSession, playerId: UUID, amount: Int) -> Result<Transaction, TransferError> {
        guard amount > 0 else {
            return .failure(.invalidAmount)
        }
        guard let index = session.players.firstIndex(where: { $0.id == playerId }) else {
            return .failure(.invalidPlayer)
        }
        guard session.players[index].chipCount >= amount else {
            return .failure(.insufficientChips)
        }

        session.players[index].chipCount -= amount
        session.pot += amount

        let transaction = Transaction.bet(playerId: playerId, amount: amount)
        session.transactions.append(transaction)
        return .success(transaction)
    }

    func collectPot(session: inout GameSession, winnerId: UUID) -> Result<Transaction, TransferError> {
        let pot = session.pot
        guard pot > 0 else {
            return .failure(.emptyPot)
        }
        guard let index = session.players.firstIndex(where: { $0.id == winnerId }) else {
            return .failure(.invalidPlayer)
        }

        session.players[index].chipCount += pot
        session.pot = 0

        let transaction = Transaction.potWin(playerId: winnerId, amount: pot)
        session.transactions.append(transaction)
        return .success(transaction)
    }

    func currentRoundBet(session: GameSession, playerId: UUID) -> Int {
        session.transactions
            .reversed()
            .prefix { $0.type != .potWin }
            .filter { $0.type == .bet && $0.fromPlayerId == playerId }
            .reduce(0) { $0 + $1.amount }
    }
}

File: ChipTallyAOS/app/src/main/java/com/chiptally/domain/usecase/PotUseCase.kt
package com.chiptally.domain.usecase

import com.chiptally.domain.model.GameSession
import com.chiptally.domain.model.Transaction
import com.chiptally.domain.model.TransactionType
import com.chiptally.domain.model.TransferError

/**
 * 보드 가운데 팟에 칩을 걸고, 승자가 전액을 가져가는 흐름.
 *
 * '이번 판'은 마지막 [TransactionType.POT_WIN] 이후를 뜻한다. 별도 상태를 두지 않고
 * 거래 내역에서 계산하므로 저장/복원과 자동으로 맞아떨어진다.
 */
class PotUseCase {

    fun bet(session: GameSession, playerId: String, amount: Int): Result<Transaction> {
        if (amount <= 0) {
            return Result.failure(TransferException(TransferError.InvalidAmount))
        }

        val player = session.players.find { it.id == playerId }
            ?: return Result.failure(TransferException(TransferError.InvalidPlayer))

        if (player.chipCount < amount) {
            return Result.failure(TransferException(TransferError.InsufficientChips))
        }

        player.chipCount -= amount
        session.pot += amount

        val transaction = Transaction.bet(playerId = playerId, amount = amount)
        session.transactions.add(transaction)
        return Result.success(transaction)
    }

    fun collectPot(session: GameSession, winnerId: String): Result<Transaction> {
        val pot = session.pot
        if (pot <= 0) {
            return Result.failure(TransferException(TransferError.EmptyPot))
        }

        val winner = session.players.find { it.id == winnerId }
            ?: return Result.failure(TransferException(TransferError.InvalidPlayer))

        winner.chipCount += pot
        session.pot = 0

        val transaction = Transaction.potWin(playerId = winnerId, amount = pot)
        session.transactions.add(transaction)
        return Result.success(transaction)
    }

    /** 이번 판에 [playerId] 가 팟에 넣은 총액. '묻고 더블로 가!' 의 기준이 된다. */
    fun currentRoundBet(session: GameSession, playerId: String): Int =
        session.transactions
            .takeLastWhile { it.type != TransactionType.POT_WIN }
            .filter { it.type == TransactionType.BET && it.fromPlayerId == playerId }
            .sumOf { it.amount }
}
